### PR TITLE
Use string rules

### PR DESCRIPTION
**Pull request by pjdufour-truss**
This PR adds a `allowed_hosts` and `blocked_path_prefixes` variables that are used in byte match rules (aka string rules) the replace the functionality of the existing regex rules.  Additionally, this PR adds a `rules` variable that would allow regex rules, if they are indeed needed.

### CODE REVIEW

**Comment by pjdufour-truss** [comment]
> _No description provided._

File: README.md [hunk 1]
```diff
@@ -24,8 +24,8 @@ module "waf" {
 
   alb_arn                             = "${module.alb_web_containers.alb_arn}"
   associate_alb                       = true
-  ip_rate_limit                       = 2000
   ip_sets                             = [aws_wafregional_ipset.global.id]
+  rate_based_rules                    = [aws_wafregional_rate_based_rule.ipratelimit.id]
   regex_host_allow_pattern_strings    = var.waf_regex_host_allow_pattern_strings
```
Comment (dynamike) [inline README.md]: Remove deprecated `regex_host_allow_pattern_strings` and `regex_path_disallow_pattern_strings`

File: README.md [hunk 2]
```diff
@@ -24,8 +24,8 @@ module "waf" {
 
   alb_arn                             = "${module.alb_web_containers.alb_arn}"
   associate_alb                       = true
-  ip_rate_limit                       = 2000
   ip_sets                             = [aws_wafregional_ipset.global.id]
+  rate_based_rules                    = [aws_wafregional_rate_based_rule.ipratelimit.id]
```
Comment (dynamike) [inline README.md]: can you include `aws_wafregional_rate_based_rule.ipratelimit` in this example so it's clear how it should be used?

File: README.md [hunk 3]
```diff
@@ -24,8 +24,8 @@ module "waf" {
```
Comment (dynamike) [inline README.md]: The example is missing required `blocked_path_prefixes` and  `allowed_hosts`

File: README.md [hunk 4]
```diff
@@ -82,3 +85,25 @@ Version `2.0.0` removes the `environment` variable and adds `web_acl_metric_name
 Version `2.0.0` replaces the `ip_set` variable with a `ip_sets` list variable, which accepts a list of `aws_wafregional_ipset` ids.  This variable allows the Web ACL to pull from multiple lists of blocked ip addresses, such that you can combine a global blocked list, and application-specific lists.  For example: `ip_sets = [resource.aws_wafregional_ipset.global.id, resource.aws_wafregional_ipset.helloworld.id]`.
 
 During the initial upgrade to `2.0.0`, and if you add additional dynamic rules, you'll need to delete your web ACLs, as terraform cannot properly handle peer dependencies between Rules and Web ACLs.  For convenience, you can use the `delete-web-acl` script in the scripts folder.  For example: `scripts/delete-web-acl WEB_ACL_ID`.  Once the Web ACL is deleted use terraform apply to recreate the Web ACL and associate with your resources as you had before.  Deleting a Web ACL does not delete any associated resources, such as Application Load Balancers; however, it will leave the resources temporarily unprotected.
+
+### 2.0.0 to 2.1.0
```
Comment (dynamike) [inline README.md]: It looks like we order new upgrade path steps above older upgrade paths see
https://github.com/trussworks/terraform-aws-logs#upgrade-paths

Comment (dynamike) [inline README.md]: It looks like we order new upgrade paths above older ones See
https://github.com/trussworks/terraform-aws-logs#upgrade-paths

File: main.tf [hunk 5]
```diff
@@ -144,45 +143,61 @@ resource "aws_wafregional_web_acl" "wafacl" {
...
-  rule {
-    type     = "REGULAR"
-    rule_id  = aws_wafregional_rule.regex_host.id
-    priority = 1 + length(aws_wafregional_rule.ips.*.id) + 1 + (length(var.wafregional_rule_f5_id) > 0 ? 1 : 0) + 1
+  dynamic "rule" {
+    for_each = var.rules
+    content {
+      type     = "REGULAR"
+      rule_id  = rule.value
+      priority = 1 + length(aws_wafregional_rule.ips.*.id) + 1 + 1 + length(var.rate_based_rules) + (length(var.wafregional_rule_f5_id) > 0 ? 1 : 0) + rule.key
```
Comment (cblkwell) [inline main.tf]: I love how this looks like an escalating Jenga tower of priorities here. :)

File: variables.tf [hunk 6]
```diff
@@ -9,26 +9,32 @@ variable "alb_arn" {
...
+  description = "List of WAF Rate-based rules."
+  type        = "list"
+  default     = []
 }
 
-variable "ip_rate_limit" {
-  description = "The rate limit for IPs matching with a 5 minute window."
-  type        = number
-  default     = 2000
+variable "rules" {
```
Comment (dynamike) [inline variables.tf]: How are `rules` different than `rate_based_rules`?

File: variables.tf [hunk 7]
```diff
@@ -9,26 +9,32 @@ variable "alb_arn" {
...
-  description = "The list of URI paths to block using the WAF."
-  type        = list(string)
-}
-
-variable "regex_host_allow_pattern_strings" {
-  description = "The list of hosts to allow using the WAF (as found in HTTP Header)."
-  type        = list(string)
+variable "rate_based_rules" {
+  description = "List of WAF Rate-based rules."
+  type        = "list"
```
Comment (dynamike) [inline variables.tf]: Is this a list of object or strings?

File: variables.tf [hunk 8]
```diff
@@ -9,26 +9,32 @@ variable "alb_arn" {
...
+  default     = []
 }
 
-variable "ip_rate_limit" {
-  description = "The rate limit for IPs matching with a 5 minute window."
-  type        = number
-  default     = 2000
+variable "rules" {
+  description = "List of WAF rules."
+  type        = "list"
```
Comment (dynamike) [inline variables.tf]: Similar to above, is this a list of object or strings?

File: README.md [hunk 9]
```diff
@@ -40,11 +40,12 @@ module "waf" {
...
 | alb\_arn | ARN of the Application Load Balancer (ALB) to be associated with the Web Application Firewall (WAF) Access Control List (ACL). | string | n/a | yes |
+| allowed\_hosts | The list of allowed host names as specified in HOST header. | list(string) | n/a | yes |
 | associate\_alb | Whether to associate an Application Load Balancer (ALB) with an Web Application Firewall (WAF) Access Control List (ACL). | bool | `"false"` | no |
-| ip\_rate\_limit | The rate limit for IPs matching with a 5 minute window. | number | `"2000"` | no |
+| blocked\_path\_prefixes | The list of URI path prefixes to block using the WAF. | list(string) | n/a | yes |
 | ip\_sets | List of sets of IP addresses to block. | list(string) | `[]` | no |
-| regex\_host\_allow\_pattern\_strings | The list of hosts to allow using the WAF (as found in HTTP Header). | list(string) | n/a | yes |
-| regex\_path\_disallow\_pattern\_strings | The list of URI paths to block using the WAF. | list(string) | n/a | yes |
+| rate\_based\_rules | List of WAF Rate-based rules. | list | `[]` | no |
+| rules | List of WAF rules. | list | `[]` | no |
```
[the remark this answers is not in the document]
Comment (pjdufour-truss) [reply README.md]: Right, I should clarify that.  They are mutually exclusive.

**Pull request merged**